presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterThroughRowNumber.java

```diff
@@ -0,0 +1,186 @@
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.RowNumberNode;
+import io.prestosql.sql.planner.plan.ValuesNode;
+import io.prestosql.sql.tree.BooleanLiteral;
+import io.prestosql.sql.tree.Expression;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static io.prestosql.spi.predicate.Range.range;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.DomainTranslator.fromPredicate;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.rowNumber;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterThroughRowNumber
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter().with(source().matching(rowNumber()));
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+
+    public PushdownFilterThroughRowNumber(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+
+        PlanNode source = getOnlyElement(lookup.resolveGroup(node.getSource()).collect(Collectors.toList()));
+
+        TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+        if (source instanceof RowNumberNode) {
+            Symbol rowNumberSymbol = ((RowNumberNode) source).getRowNumberSymbol();
+            OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+            if (upperBound.isPresent()) {
+                if (upperBound.getAsInt() <= 0) {
+                    return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+                }
+                RowNumberNode rowNumberNode = (RowNumberNode) source;
+                if (rowNumberNode.getMaxRowCountPerPartition().isPresent()) {
+                    if (rowNumberNode.getMaxRowCountPerPartition().get().equals(upperBound.getAsInt())) {
+                        return Result.empty();
+                    }
+                }
```

Comment:
This doesn't seem to match WindowFilterPushdown.visitFilter

Reply:
This was necessary to avoid the infinite loop. Without this condition,`PushdownFilterThroughRowNumber` always the non-empty result even if the row number node is already rewritten. It caused the timeout of the iterative-optimizer in several test cases.